File: Problem#27.py
#ran on leetcode

class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        result=[]
        
        n=len(matrix)     #no of row
        m=len(matrix[0])  #no of col
        
        #defining pointers
        left=0
        right=m-1
        top=0
        bottom=n-1

        # edge condition check
        if(matrix==0 or len(matrix)==0):
            return result
            #print "Null Array"

        #
        while(left<=right and top<=bottom):
            #case 1: top row:l->r
            for i in range(left,right+1):
                result.append(matrix[top][i])
            top+=1
            #case 2: right col:t->b
            for i in range(top,bottom+1):
                result.append(matrix[i][right])
            right-=1
            
            #case 3: bottom row:r->l; check t<=b
            if(top<=bottom):
                for i in range(right,left-1,-1): #note left-1
                    result.append(matrix[bottom][i])
                bottom-=1
            #case 4: left col:b->t; check l<=r
            if(left<=right):
                for i in range(bottom,top-1,-1):  #note top-1
                    result.append(matrix[i][left])
                left+=1
        return result
    

######################
#Complexity
#Time:O(mn)  # we are going through every element
#Space:O(1)  # we are not using any extra space
######################




# -*- coding: utf-8 -*-
"""
Created on Sat Dec  7 21:08:21 2019

@author: gunja
"""

#spiral array proble
matrix=[[11,2,333],[41,55,66],[7,88,99]]
#matrix=[[1,2,3],[4,5,6],[7,8,9]]
result=[]

m=len(matrix[0])  # no of col
n=len(matrix)     #no of row

#defining pointers
left=0
right=m-1
top=0
bottom=n-1

# edge condition check
if(matrix==0 or len(matrix)==0):
    #return result
    print "Null Array"
    
#
while(left<=right and top<=bottom):
    #case 1: top row:l->r
    for i in range(left,right+1):
        result.append(matrix[top][i])
    top+=1
    #case 2: right col:t->b
    for i in range(top,bottom+1):
        result.append(matrix[i][right])
    right-=1
    #case 3: bottom row:r->l; check t<=b
    print top
    print bottom
    print left
    print right
    if(top<=bottom):
        for i in range(right,left-1,-1): #note left-1
            result.append(matrix[bottom][i])
        bottom-=1
    #case 4: left col:b->t; check l<=r
    if(left<=right):
        for i in range(bottom,top-1,-1):  #note top-1
            result.append(matrix[i][left])
        left+=1


print result

File: Problem#27.java
// Product of Array Except Self

// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english : For every element in the array, we have to find the products of all the elements except itself, in other words, we have to find the product of all the elements towards its left and right. For this, we create 2 arrays. First array will contain the product of all the elements towards its left and the second array will contain the product of all the elements towards the right of the element in consideration from the parent array. Now that we have left and the right products, all that we have to do is to multiply matching indices from both the arrays. To reduce the space complexity, instead of creating a second new array, we can just multiply the running product with the first array itself and output it.

// Your code here along with comments explaining your approach

class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] result = new int[nums.length];
        if(nums == null || nums.length == 0) // Check if the array itself is not initialized or if there are no elements in the array
            return result;
        int rProd = 1; 
        result[0] = rProd; // Running left product of the first element will be 1 as there is no element to the left of the element at the 0th index
        for(int i = 1; i <nums.length; i++){ // Iterate through the array, calculate the running product and update the result array
            rProd *= nums[i-1];
            result[i] = rProd;
        }
        rProd = 1; // Running right product of the last element will be 1 as there is no element to the right of the element at the last index
        for(int i = nums.length - 2; i >= 0; i--){ // Iterate through the array from backwards starting from the second last element, as the running product to multiply the last element with will be 1, calculate the running product and update the array
            rProd *= nums[i+1];
            result[i] *= rProd; // Update the same array (output array) itself to save space
        }
        return result;
    }
}